libraries/AC_AttitudeControl/AC_AttitudeControl.cpp

```diff
@@ -97,321 +97,412 @@ const AP_Param::GroupInfo AC_AttitudeControl::var_info[] = {
+    Quaternion thrust_vec_correction_quat;
+    thrust_vec_correction_quat.from_axis_angle(thrust_vec_cross, thrust_vec_dot);
+    thrust_vec_correction_quat = att_from_quat.inverse()*thrust_vec_correction_quat*att_from_quat;
+
+    // calculate the remaining rotation required after thrust vector is rotated
+    Quaternion heading_quat = thrust_vec_correction_quat.inverse()*att_from_quat.inverse()*att_to_quat;
+
+    Vector3f rotation;
+    thrust_vec_correction_quat.to_axis_angle(rotation);
+    att_diff_angle.x = rotation.x;
+    att_diff_angle.y = rotation.y;
+
+    heading_quat.to_axis_angle(rotation);
+    att_diff_angle.z = rotation.z;
+    if(fabsf(att_diff_angle.z) > AC_ATTITUDE_ACCEL_Y_CONTROLLER_MAX_RADSS/_p_angle_yaw.kP()){
+        att_diff_angle.z = constrain_float(wrap_PI(att_diff_angle.z), -radians(30.0f), radians(30.0f));
+        heading_quat.from_axis_angle(Vector3f(0.0f,0.0f,att_diff_angle.z));
+        att_to_quat = att_from_quat*thrust_vec_correction_quat*heading_quat;
+    }
 }
 
-void AC_AttitudeControl::attitude_controller_run_euler(const Vector3f& att_target_euler_rad, const Vector3f& att_target_ang_vel_rads)
+// calculates the velocity correction from an angle error. The angular velocity has acceleration and
+// deceleration limits including basic jerk limiting using smoothing_gain
+float AC_AttitudeControl::input_shaping_angle(float error_angle, float smoothing_gain, float accel_max, float target_ang_vel)
 {
-    // Compute quaternion target attitude
-    Quaternion att_target_quat;
-    att_target_quat.from_euler(att_target_euler_rad.x, att_target_euler_rad.y, att_target_euler_rad.z);
+    error_angle = wrap_PI(error_angle);
+    // Calculate the velocity as error approaches zero with acceleration limited by accel_max_radss
+    float ang_vel = sqrt_controller(error_angle, smoothing_gain, accel_max);
 
-    // Call quaternion attitude controller
-    attitude_controller_run_quat(att_target_quat, att_target_ang_vel_rads);
+    // Acceleration is limited directly to smooth the beginning of the curve.
+    float delta_ang_vel = accel_max * _dt;
+    return constrain_float(ang_vel, target_ang_vel-delta_ang_vel, target_ang_vel+delta_ang_vel);
 }
 
-void AC_AttitudeControl::attitude_controller_run_quat(const Quaternion& att_target_quat, const Vector3f& att_target_ang_vel_rads)
+// limits the acceleration and deceleration of a velocity request
+float AC_AttitudeControl::input_shaping_ang_vel(float target_ang_vel, float desired_ang_vel, float accel_max)
 {
-    // Update euler attitude target and angular velocity target
-    att_target_quat.to_euler(_att_target_euler_rad.x,_att_target_euler_rad.y,_att_target_euler_rad.z);
-    _att_target_ang_vel_rads = att_target_ang_vel_rads;
-
-    // Retrieve quaternion vehicle attitude
-    // TODO add _ahrs.get_quaternion()
-    Quaternion att_vehicle_quat;
-    att_vehicle_quat.from_rotation_matrix(_ahrs.get_rotation_body_to_ned());
+    if ((accel_max > 0.0f) && !is_zero(accel_max)) {
```

Comment:
My point was that if it is already bigger than zero, then it isn't zero, so there wouldn't be a reason for second part of the if. But since this is a float I'm not 100% sure this is true, that's why I asked if there was any reason - you could have caught something during testing.
 
You are obviously the expert in this, but, as a general rule, I would prefer to get a negative accel max and convert it to a positive than just ignore it. What's the worst that can happen in that case? I mean, wouldn't we just be limiting the acceleration instead of no limit? No limit seems more dangerous to me.